File: typescript/tests/commerce_board.test.ts
// commerce_board.test.ts
// Verifies the CircleAI.Commerce port: customers, orders (newest-first), line
// items (insertion order), status update, and lifetime value.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryCommerceBoard,
  CommerceDomainContext,
  commerceCustomer,
  commerceOrder,
  commerceLineItem,
} from "../src/commerce/index";

describe("InMemoryCommerceBoard", () => {
  it("adds and retrieves customers (email may be null)", () => {
    const b = new InMemoryCommerceBoard();
    b.addCustomer(commerceCustomer("c1", "Ada", null, new Date("2026-01-01T00:00:00Z")));
    assert.equal(b.getCustomer("c1")?.email, null);
    assert.equal(b.getCustomer("nope"), undefined);
  });

  it("lists orders newest-first and computes lifetime value", () => {
    const b = new InMemoryCommerceBoard();
    b.place(commerceOrder("o1", "c1", 100, "ZAR", "Paid", new Date("2026-01-01T00:00:00Z")));
    b.place(commerceOrder("o2", "c1", 250, "ZAR", "Paid", new Date("2026-06-01T00:00:00Z")));
    b.place(commerceOrder("o3", "c2", 999, "ZAR", "Paid", new Date("2026-05-01T00:00:00Z")));
    assert.deepEqual(
      b.ordersFor("c1").map((o) => o.orderId),
      ["o2", "o1"],
    );
    assert.equal(b.lifetimeValue("c1"), 350);
    assert.equal(b.lifetimeValue("unknown"), 0);
  });

  it("line items preserve insertion order and filter by order", () => {
    const b = new InMemoryCommerceBoard();
    b.addLine(commerceLineItem("L1", "o1", "SKU-A", 2, 50));
    b.addLine(commerceLineItem("L2", "o2", "SKU-B", 1, 10));
    b.addLine(commerceLineItem("L3", "o1", "SKU-C", 3, 5));
    assert.deepEqual(
      b.linesFor("o1").map((l) => l.lineId),
      ["L1", "L3"],
    );
  });

  it("updateStatus mutates the order; unknown throws", () => {
    const b = new InMemoryCommerceBoard();
    b.place(commerceOrder("o1", "c1", 100, "ZAR", "Pending", new Date("2026-01-01T00:00:00Z")));
    b.updateStatus("o1", "Shipped");
    assert.equal(b.ordersFor("c1")[0].status, "Shipped");
    assert.throws(() => b.updateStatus("ghost", "X"), /Unknown order ghost/);
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(CommerceDomainContext.systemPromptSnippet.includes("[DOMAIN: Commerce]"));
    assert.deepEqual(CommerceDomainContext.complianceFlags, ["POPIA", "Consumer_Protection_Act", "GDPR_aware"]);
    assert.deepEqual(CommerceDomainContext.suggestedTools, ["inventory", "pricing_engine", "order_management", "analytics"]);
  });
});

File: go/tests/commerce_board_test.go
// commerce_board_test.go
//
// Verifies the CircleAI.Commerce port (commerce_board.go): customer add/get,
// order place/status, line items (insertion order), orders-for ordering (desc),
// and lifetime value summation.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestCommerce_CustomerAddGet(t *testing.T) {
	b := circleai.NewInMemoryCommerceBoard()
	b.AddCustomer(circleai.CommerceCustomer{CustomerId: "cust1", Name: "Ada", Email: strptr("ada@x.io"), CreatedUtc: time.Now().UTC()})
	got, ok := b.GetCustomer("cust1")
	if !ok || got.Name != "Ada" || got.Email == nil || *got.Email != "ada@x.io" {
		t.Fatalf("get customer = %+v ok=%v", got, ok)
	}
	// Nil email preserved.
	b.AddCustomer(circleai.CommerceCustomer{CustomerId: "cust2", Name: "NoMail", Email: nil})
	g2, _ := b.GetCustomer("cust2")
	if g2.Email != nil {
		t.Fatalf("nil email should stay nil")
	}
}

func TestCommerce_OrdersLinesAndStatus(t *testing.T) {
	b := circleai.NewInMemoryCommerceBoard()
	base := time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)
	b.Place(circleai.CommerceOrder{OrderId: "o1", CustomerId: "c1", Total: circleai.DecimalFromInt(100), Currency: "ZAR", Status: "New", AtUtc: base})
	b.Place(circleai.CommerceOrder{OrderId: "o2", CustomerId: "c1", Total: circleai.DecimalFromInt(250), Currency: "ZAR", Status: "New", AtUtc: base.Add(48 * time.Hour)})
	b.Place(circleai.CommerceOrder{OrderId: "o3", CustomerId: "c1", Total: circleai.DecimalFromInt(50), Currency: "ZAR", Status: "New", AtUtc: base.Add(24 * time.Hour)})

	orders := b.OrdersFor("c1")
	if len(orders) != 3 || orders[0].OrderId != "o2" || orders[1].OrderId != "o3" || orders[2].OrderId != "o1" {
		t.Fatalf("orders desc failed: %+v", orders)
	}
	if err := b.UpdateStatus("o1", "Shipped"); err != nil {
		t.Fatalf("update status: %v", err)
	}
	if err := b.UpdateStatus("ghost", "X"); err == nil {
		t.Fatalf("unknown order status must error")
	}

	// Lines preserve insertion order and filter by order.
	b.AddLine(circleai.CommerceLineItem{LineId: "L1", OrderId: "o1", Sku: "A", Quantity: 2, UnitPrice: circleai.DecimalFromInt(10)})
	b.AddLine(circleai.CommerceLineItem{LineId: "L2", OrderId: "o2", Sku: "B", Quantity: 1, UnitPrice: circleai.DecimalFromInt(5)})
	b.AddLine(circleai.CommerceLineItem{LineId: "L3", OrderId: "o1", Sku: "C", Quantity: 3, UnitPrice: circleai.DecimalFromInt(7)})
	lines := b.LinesFor("o1")
	if len(lines) != 2 || lines[0].LineId != "L1" || lines[1].LineId != "L3" {
		t.Fatalf("lines insertion order failed: %+v", lines)
	}

	// Lifetime value = 100 + 250 + 50 = 400.
	if ltv := b.LifetimeValue("c1"); !ltv.Equal(circleai.DecimalFromInt(400)) {
		t.Fatalf("ltv = %s, want 400", ltv)
	}
	if ltv := b.LifetimeValue("nobody"); !ltv.IsZero() {
		t.Fatalf("ltv for unknown customer = %s, want 0", ltv)
	}
}
